torch/_tensor_str.py

```diff
@@ -138,178 +139,73 @@ def _number_format(tensor, min_sz=-1):
-            for row in self:
-                strt += __repr_row(row, indent, fmt, scale, sz, n)
-        elif has_vdots and not has_hdots:
-            vdotfmt = u"{:^" + \
-                str(len(__repr_row(self[0], '', fmt, scale, sz))) + \
-                "}\n"
-            for row in self[:n]:
-                strt += __repr_row(row, indent, fmt, scale, sz)
-            strt += vdotfmt.format(u'\u22EE')
-            for row in self[-n:]:
-                strt += __repr_row(row, indent, fmt, scale, sz)
-        else:
-            for row in self:
-                strt += __repr_row(row, indent, fmt, scale, sz)
-    return strt
-
-
-def _vector_str(self):
-    fmt, scale, sz = _number_format(self)
-    strt = ''
-    ident = ''
-    n = PRINT_OPTS.edgeitems
-    dotfmt = u"{:^" + str(sz) + "}\n"
-    if scale != 1:
-        strt += SCALE_FORMAT.format(scale)
-        ident = ' '
-    if self.numel() < PRINT_OPTS.threshold:
-        return (strt +
-                '\n'.join(ident + fmt.format(val.item() / scale) for val in self) +
-                '\n')
+        data = [fmt.format(val.item() / scale) for val in self]
+
+    data_lines = [data[i:i + elements_per_line] for i in range(0, len(data), elements_per_line)]
+    lines = [','.join(line) for line in data_lines]
+    return '[' + (',' + '\n' + ' ' * (indent + 1)).join(lines) + ']'
+
+
+def _tensor_str(self, indent, fmt, scale, sz):
+    dim = self.dim()
+
+    if dim == 0:
+        return _scalar_str(self, fmt, scale)
+    if dim == 1:
+        return _vector_str(self, indent, fmt, scale, sz)
+
+    if self.size(0) > PRINT_OPTS.threshold:
+        slices = ([_tensor_str(self[i], indent + 1, fmt, scale, sz) for i in range(0, PRINT_OPTS.edgeitems)] +
+                  ['...'] +
+                  [_tensor_str(self[i], indent + 1, fmt, scale, sz) for i in range(len(self) - PRINT_OPTS.edgeitems,
+                                                                                   len(self))])
     else:
-        return (strt +
-                '\n'.join(ident + fmt.format(val.item() / scale) for val in self[:n]) +
-                '\n' + (ident + dotfmt.format(u"\u22EE")) +
-                '\n'.join(ident + fmt.format(val.item() / scale) for val in self[-n:]) +
-                '\n')
+        slices = [_tensor_str(self[i], indent + 1, fmt, scale, sz) for i in range(0, self.size(0))]
+
+    tensor_str = (',' + '\n' * (dim - 1) + ' ' * (indent + 1)).join(slices)
+    return '[' + tensor_str + ']'
 
 
-def _str(self, include_footer=True):
+def _str(self):
     if self.is_sparse:
         size_str = str(tuple(self.shape)).replace(' ', '')
         return '{} of size {} with indices:\n{}and values:\n{}'.format(
             self.type(), size_str, self._indices(), self._values())
 
-    empty = self.numel() == 0
-    dim = self.dim()
+    type_str = 'tensor'
+    prefix = type_str + '('
+    indent = len(prefix)
+
+    suffix = ')'
+    if str(self.device.type) != 'cpu':
```

Comment:
I think these are trying to follow the rule of "make these implicit if torch.tensor would return you something with that device/dtype automatically", but there are issues here:
1) The device type will be cuda by default if torch.set_default_tensor_type is a cuda type.  Otherwise, it will be a CPU type.
2) Similar thing for implicit dtype: torch.set_default_tensor_type must be a floating point type (half, float, double), but this assumes it's always float32.

Also, why don't you print the actual device (you could make it implicit if you prefer if the default tensor type is a cuda type and the current device is the device of the tensor, but you never print it?  That doesn't follow the above rule either.

Reply:
Also, the integeral-vs-floating printing is wrong, for example:
```
>>> a=torch.tensor((1., 2, 3))
>>> a
tensor([ 1, 2, 3])
>>> b=torch.tensor([ 1, 2, 3])
>>> a.dtype == b.dtype
False
```